File: Data Structures & Algorithms/search-2d-matrix/submission-1.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int l = 0;
        int r = matrix.length-1;
        int mid = 0;
        while(l<=r){
            mid = (l+r)/2;
            if(target>= matrix[mid][0] && target <=matrix[mid][matrix[0].length-1]){
                break;
            }
            
            if(target>=matrix[mid][0]) {
                l = mid + 1;
            }
            else {
                r = mid - 1;
            }
        }

        l = 0;
        r = matrix[0].length-1;
        int index = mid;
        while(l<=r){
            mid = (l+r)/2;
            System.out.println(l + " " + r);
            if(target == matrix[index][mid]){
                return true;
            }
            if(target > matrix[index][mid]){
                l = mid + 1;
            }
            else {
                r = mid - 1;
            }

        }
        return false;
    }
}

File: Data Structures & Algorithms/target-sum/submission-1.py
import functools
class Solution:

    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        n = len(nums)
        @functools.cache
        def dp(i, j):
            print(j)
            #i index, j curr sum
            if i >= n:
                if j == 0:
                    return 1
                return 0
            return dp(i+1, j+nums[i]) + dp(i+1, j-nums[i])
        return dp(0, target)
            